server/coordinator/scheduler/nodepicker/hash/consistent_uniform.go

```diff
@@ -220,24 +281,96 @@ func (c *ConsistentUniformHash) initializeVirtualNodes(members []Member) {
 	})
 }
 
-// LoadDistribution exposes load distribution of members.
-func (c *ConsistentUniformHash) LoadDistribution() map[string]float64 {
-	// Create a thread-safe copy
-	res := make(map[string]float64)
-	for member, load := range c.memLoads {
-		res[member] = load
+func (c *ConsistentUniformHash) ensureAffinity() {
+	offloadedMems := make(map[string]struct{}, len(c.config.PartitionAffinities))
+
+	for _, affinity := range c.config.PartitionAffinities {
+		partID := affinity.PartitionID
+		vNodeIdx := c.partitionDist[partID]
+		vNode := c.sortedRing[vNodeIdx]
+		mem, ok := c.nodeToMems[vNode]
+		assert.Assert(ok)
+		offloadedMems[mem.String()] = struct{}{}
+
+		allowedLoad := int(affinity.NumAllowedOtherPartitions) + 1
+		memPartIDs, ok := c.memPartitions[mem.String()]
+		assert.Assert(ok)
+		memLoad := len(memPartIDs)
+		if memLoad > allowedLoad {
+			c.offloadMember(mem, memPartIDs, partID, allowedLoad, offloadedMems)
+		}
 	}
-	return res
 }
 
-// GetPartitionOwner returns the owner of the given partition.
-func (c *ConsistentUniformHash) GetPartitionOwner(partID int) Member {
-	virtualNodeIdx, ok := c.partitionDist[partID]
-	if !ok {
-		return nil
+// offloadMember tries to offload the given member by moving the its partitions to another members.
+func (c *ConsistentUniformHash) offloadMember(mem Member, memPartitions map[int]struct{}, retainedPartID, numAllowedParts int, offloadedMems map[string]struct{}) {
+	assert.Assertf(numAllowedParts >= 1, "At least the partition itself should be allowed")
+	partIDsToOffload := make([]int, 0, len(memPartitions)-numAllowedParts)
+	// The `retainedPartID` must be retained.
+	numRetainedParts := 1
+	for partID := range memPartitions {
+		if partID == retainedPartID {
+			continue
+		}
+
+		if numRetainedParts < numAllowedParts {
+			numRetainedParts++
+			continue
+		}
+
+		partIDsToOffload = append(partIDsToOffload, partID)
 	}
-	virtualNode := c.sortedRing[virtualNodeIdx]
-	mem, ok := c.nodeToMems[virtualNode]
-	assert.Assertf(ok, "member must exist for the virtual node")
-	return mem
+
+	slices.Sort(partIDsToOffload)
+	for _, partID := range partIDsToOffload {
+		c.offloadPartition(partID, mem, offloadedMems)
+	}
+}
+
+func (c *ConsistentUniformHash) offloadPartition(sourcePartID int, sourceMem Member, blackedMembers map[string]struct{}) {
+	// Ensure all members' load smaller than the max load as much as possible.
+	loadUpperBound := c.numPartitions
+	for load := c.maxLoad; load < int(loadUpperBound); load++ {
+		if done := c.offloadPartitionWithAllowedLoad(sourcePartID, sourceMem, load, blackedMembers); done {
+			break
+		}
+	}
+}
+
+func (c *ConsistentUniformHash) offloadPartitionWithAllowedLoad(sourcePartID int, sourceMem Member, allowedMaxLoad int, blackedMembers map[string]struct{}) bool {
+	vNodeIdx := c.partitionDist[sourcePartID]
+	// Skip the first member which must not be the target to move.
```

Comment:
I can't understand, why do you need to skip the first member?

Reply:
Because the `sourcePartitionID` belongs to the firstMemer (`sourceMem`).